resourcestore/fileresourcestore.go

```diff
@@ -0,0 +1,468 @@
+package resourcestore
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/cluster"
+	"github.com/weaveworks/flux/image"
+	"github.com/weaveworks/flux/policy"
+	"github.com/weaveworks/flux/resource"
+)
+
+type resourceWithOrigin struct {
+	resource   resource.Resource
+	configFile *ConfigFile // only set if the resource came from a configuration file
+}
+
+type fileResourceStore struct {
+	ctx             context.Context
+	manifests       cluster.Manifests
+	baseDir         string
+	rawManifestDirs []string
+	configFiles     []*ConfigFile
+	resourcesByID   map[string]resourceWithOrigin
+	sync.RWMutex
+}
+
+var _ ResourceStore = &fileResourceStore{}
+
+func NewFileResourceStore(ctx context.Context, baseDir string, targetPaths []string, enableManifestGeneration bool,
+	manifests cluster.Manifests) (*fileResourceStore, error) {
+	var (
+		err             error
+		configFiles     []*ConfigFile
+		rawManifestDirs []string
+	)
+
+	rawManifestDirs = targetPaths
+	if enableManifestGeneration {
+		configFiles, rawManifestDirs, err = splitConfigFilesAndRawManifestPaths(baseDir, targetPaths)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	result := &fileResourceStore{
+		ctx:             ctx,
+		manifests:       manifests,
+		baseDir:         baseDir,
+		rawManifestDirs: rawManifestDirs,
+		configFiles:     configFiles,
+	}
+	return result, nil
+}
+
+func splitConfigFilesAndRawManifestPaths(baseDir string, paths []string) ([]*ConfigFile, []string, error) {
+	var (
+		configFiles      []*ConfigFile
+		rawManifestPaths []string
+	)
+
+	for _, path := range paths {
+		relPath, err := filepath.Rel(baseDir, path)
+		if err != nil {
+			return nil, nil, err
+		}
+		configFilePath, workingDirPath, err := findConfigFilePaths(baseDir, path)
+		if err != nil {
+			if err == configFileNotFoundErr {
+				rawManifestPaths = append(rawManifestPaths, path)
+				continue
+			}
+			return nil, nil, fmt.Errorf("error when searching config files for path %q: %s", relPath, err)
+		}
+		cf, err := NewConfigFile(configFilePath, workingDirPath)
+		if err != nil {
+			relConfigFilePath, relErr := filepath.Rel(baseDir, configFilePath)
+			if err != nil {
+				return nil, nil, relErr
+			}
+			return nil, nil, fmt.Errorf("cannot parse config file %q: %s", relConfigFilePath, err)
+		}
+		configFiles = append(configFiles, cf)
+	}
+
+	return configFiles, rawManifestPaths, nil
+}
+
+var configFileNotFoundErr = fmt.Errorf("config file not found")
+
+func findConfigFilePaths(baseDir string, initialPath string) (string, string, error) {
```

Comment:
Yes, it could be. However:
1. I initially made it so that we only searched for config files in the directory itself, but it didn't allow sharing config files for different directories.
2. I think that people will rarely mix config files and raw manifests, but we will see.